server/py/services/api/crud/alert_activation.py

```diff
@@ -15,54 +15,117 @@
 
 from datetime import datetime, timedelta
 
+from sqlalchemy.orm import Session
+
 import mlrun.common.schemas.alert
 import mlrun.utils.singleton
+import services.api.db.sqldb.db
 
 
 class AlertActivation(
     metaclass=mlrun.utils.singleton.Singleton,
 ):
-    @staticmethod
-    def get_partition_info(
-        partition_interval: str,
-        partition_datetime: datetime,
-    ) -> tuple[str, str, str]:
+    def create_and_drop_partitions(
+        self,
+        session: Session,
+        retention_days: int,
+    ) -> None:
+        """
+        Creates partitions for the future based on the specified retention days
+        and drops old partitions that are older than the retention period.
+
+        :param session: SQLAlchemy session for database operations.
+        :param retention_days: The number of days to retain partitions.
         """
-        Generates partition details for a specified interval and datetime.
 
-        :param partition_interval: The partitioning interval type, e.g., "DAY", "MONTH", or "YEARWEEK".
-        :param partition_datetime: The datetime used for generating partition details.
+        # Retrieve the partition expression and interval from the database.
+        partition_expression, partition_interval = (
+            self.get_partition_expression_and_interval(session)
+        )
+
+        # Ensure partitions for double the retention time.
+        partition_number = partition_interval.get_number_of_partitions(
+            days=2 * retention_days
+        )
+
+        # Create the calculated number of partitions.
+        self.create_partitions(session, partition_number)
+
+        # Drop old partitions that exceed the retention period.
+        self.drop_partitions(session, retention_days)
+
+    def create_partitions(
+        self,
+        session: Session,
+        partition_number: int,
+    ):
+        # Retrieve the partition function from the database
+        partition_expression, partition_interval = (
+            self.get_partition_expression_and_interval(session)
+        )
+
+        partitioning_information_list = []
+        current_datetime = datetime.now()
+
+        for _ in range(partition_number):
+            partitioning_information_list.append(
+                partition_interval.get_partition_info(current_datetime)
+            )
+
+            # Move to the next interval based on the partition_interval
+            current_datetime = mlrun.common.schemas.alert.PartitionInterval(
+                partition_interval
+            ).get_next_partition_time(current_datetime=current_datetime)
```

Comment:
can you optimize the partition creation process by precomputing the partition boundaries and avoiding multiple calls to get_next_partition_time within the loop ? this can reduce the overhead of repeated calls to get_next_partition_time and make the loop more efficient.